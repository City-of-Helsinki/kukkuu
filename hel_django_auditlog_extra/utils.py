from typing import List, Set

from auditlog.registry import auditlog
from django.apps import apps
from django.conf import settings
from django.core.exceptions import ImproperlyConfigured
from django.db.models.base import ModelBase


class AuditLogConfigurationHelper:
    """
    A helper class for managing audit log configuration in your Django project.

    This class provides methods to:

    - Retrieve all models in your project.
    - Identify models that are not explicitly configured for audit logging.
    - Raise an error if any models are not configured when
        `AUDITLOG_INCLUDE_ALL_MODELS` is enabled.

    This helps ensure that all models are either explicitly included or excluded
    from audit logging, preventing accidental omissions.
    """

    @staticmethod
    def get_app_models() -> List[ModelBase]:
        """
        Fetch all models in your Django project, including those automatically
        created by Django (like those for user accounts or sessions).
        """
        return apps.get_models(include_auto_created=True)

    @staticmethod
    def get_app_model_key(model: ModelBase) -> str:
        """
        Returns a string representation of the model,
        like 'app_label.model_name'.
        """
        return f"{model._meta.app_label}.{model._meta.model_name}"

    @staticmethod
    def get_model_classes(app_model: str) -> List[ModelBase]:
        return auditlog._get_model_classes(app_model)

    @classmethod
    def print_app_models(cls) -> None:
        """Prints all models in the project to the console."""
        for model in cls.get_app_models():
            print(cls.get_app_model_key(model))

    @classmethod
    def get_excluded_by_default(cls):
        return set(
            cls.get_model_classes(app_model)[0]
            for app_model in auditlog.DEFAULT_EXCLUDE_MODELS
        )

    @classmethod
    def get_excluded_models(cls) -> Set[ModelBase]:
        """
        Returns a set of models that are explicitly excluded from audit logging.
        """
        excluded_models: Set[ModelBase] = set()
        if (
            hasattr(settings, "AUDITLOG_INCLUDE_ALL_MODELS")
            and settings.AUDITLOG_INCLUDE_ALL_MODELS
        ):
            excluded_models = set(
                auditlog._get_exclude_models(
                    getattr(settings, "AUDITLOG_EXCLUDE_TRACKING_MODELS", [])
                )
            )

        return excluded_models | cls.get_excluded_by_default()

    @classmethod
    def get_included_models(cls) -> Set[ModelBase]:
        """Returns a set of models in included mapping for audit logging."""
        included_models = []
        models = getattr(settings, "AUDITLOG_INCLUDE_TRACKING_MODELS", [])
        for model in models:
            if isinstance(model, str):
                for model_class in cls.get_model_classes(model):
                    included_models.append(model_class)
            elif isinstance(model, dict):
                appmodel = cls.get_model_classes(model["model"])
                included_models.append(appmodel[0])
        return set(included_models)

    @classmethod
    def get_unconfigured_models(cls) -> Set[ModelBase]:
        """
        Returns a set of models that are neither registered nor excluded
        from audit logging.
        """
        excluded_models = cls.get_excluded_models()
        included_models = cls.get_included_models()
        all_models = set(cls.get_app_models())
        return all_models - excluded_models - included_models

    @classmethod
    def raise_error_if_unconfigured_models(cls) -> None:
        """
        Raises an ImproperlyConfigured exception if there are any models
        that are not explicitly configured for audit logging.
        """
        unconfigured_models = cls.get_unconfigured_models()
        if unconfigured_models:
            unconfigured_keys = [
                cls.get_app_model_key(model) for model in unconfigured_models
            ]
            raise ImproperlyConfigured(
                "AUDITLOG_INCLUDE_ALL_MODELS is enabled, requiring "
                "explicit configuration of all models. "
                "The following models ({count}) are not configured: {entries}. "
                "Please register them with `auditlog.register` "
                "or add them to `AUDITLOG_EXCLUDE_TRACKING_MODELS` "
                "in your Django settings.".format(
                    count=len(unconfigured_keys),
                    entries=", ".join(f"`{key}`" for key in unconfigured_keys),
                )
            )
