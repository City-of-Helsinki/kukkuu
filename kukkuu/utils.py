from .views import error_codes


def get_kukkuu_error_by_code(error_code):
    if not error_code:
        return None
    return next(
        (error for error, code in error_codes.items() if code == error_code), None
    )
