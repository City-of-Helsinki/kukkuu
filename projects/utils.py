def project_permission_name(permission: str) -> str:
    """
    Permission name for a project permission.

    Example:
        project_permission_name(ProjectPermission.ADMIN.value) == "projects.admin"
    """
    return f"projects.{permission}"
