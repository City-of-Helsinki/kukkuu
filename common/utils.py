import binascii
from copy import deepcopy
from functools import wraps

from django.conf import settings
from django.core.exceptions import PermissionDenied
from django.db import transaction
from graphene import Node
from graphql.execution.base import ResolveInfo
from graphql_relay import from_global_id, to_global_id

from kukkuu import __version__
from kukkuu.exceptions import DataValidationError, ObjectDoesNotExistError
from kukkuu.settings import REVISION


def update_object(obj, data):
    if not data:
        return
    for k, v in data.items():
        if v is None and not obj.__class__._meta.get_field(k).null:
            raise DataValidationError(f"{k} cannot be null.")
        setattr(obj, k, v)
    obj.save()


@transaction.atomic
def update_object_with_translations(model, model_data):
    model_data = deepcopy(model_data)
    translations_input = model_data.pop("translations", None)
    if translations_input:
        model.create_or_update_translations(translations_input)
    update_object(model, model_data)


def get_api_version():
    return " | ".join((__version__, REVISION))


def get_global_id(obj):
    return to_global_id(f"{obj.__class__.__name__}Node", obj.pk)


def get_node_id_from_global_id(global_id, expected_node_name):
    if not global_id:
        return None
    try:
        name, id = from_global_id(global_id)
    except (
        binascii.Error,
        UnicodeDecodeError,
    ):  # invalid global ID
        return None
    return id if name == expected_node_name else None


def check_can_user_administer(obj, user):
    try:
        yes_we_can = obj.can_user_administer(user)
    except AttributeError:
        raise TypeError(
            f"{obj.__class__.__name__} model does not implement can_user_administer()."
        )
    if not yes_we_can:
        raise PermissionDenied()


def get_obj_from_global_id(info, global_id, expected_obj_type):
    obj = Node.get_node_from_global_id(info, global_id)
    if not obj or type(obj) is not expected_obj_type:
        raise ObjectDoesNotExistError(
            f"{expected_obj_type.__name__} with ID {global_id} does not exist."
        )
    return obj


def get_obj_if_user_can_administer(info, global_id, expected_obj_type):
    obj = get_obj_from_global_id(info, global_id, expected_obj_type)
    check_can_user_administer(obj, info.context.user)
    return obj


def context(f):
    def decorator(func):
        def wrapper(*args, **kwargs):
            info = next(arg for arg in args if isinstance(arg, ResolveInfo))
            return func(info.context, *args, **kwargs)

        return wrapper

    return decorator


def user_passes_test(test_func):
    def decorator(f):
        @wraps(f)
        @context(f)
        def wrapper(context, *args, **kwargs):
            if test_func(context.user):
                return f(*args, **kwargs)
            raise PermissionDenied("You do not have permission to perform this action")

        return wrapper

    return decorator


# copied from https://github.com/flavors/django-graphql-jwt/blob/704f24e7ebbea0b81015ef3c1f4a302e9d432ecf/graphql_jwt/decorators.py  # noqa
login_required = user_passes_test(lambda u: u.is_authenticated)

project_user_required = user_passes_test(
    lambda u: u.is_authenticated and u.administered_projects
)


def get_translations_dict(obj, field_name):
    """
    Returns a dict of translations for a given field of a model instance.

    :param obj: The model instance
    :param field_name: The name of the field
    :return: A dict with language codes as keys and translations as values
    """

    return {
        lang_code: getattr(
            obj.translations.filter(language_code=lang_code).first(), field_name, ""
        )
        for lang_code, _ in settings.LANGUAGES
    }
