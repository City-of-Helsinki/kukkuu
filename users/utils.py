from django.contrib.auth import get_user_model
from django.core.exceptions import ValidationError
from django.core.validators import validate_email
from django_ilmoitin.utils import send_notification

from kukkuu.exceptions import InvalidEmailFormatError, VerificationTokenInvalidError
from users.notifications import NotificationType
from verification_tokens.models import VerificationToken

User = get_user_model()


def send_guardian_email_changed_notification(guardian):
    send_notification(
        guardian.email,
        NotificationType.GUARDIAN_EMAIL_CHANGED,
        context={"guardian": guardian},
        language=guardian.language,
    )


def validate_email_verification_token(user: User, verification_token_key: str):
    verification_token = VerificationToken.objects.filter(
        user=user,
        verification_type=VerificationToken.VERIFICATION_TYPE_EMAIL_VERIFICATION,
        key=verification_token_key,
    ).first()
    if not verification_token or not verification_token.is_valid():
        raise VerificationTokenInvalidError("The verification token is invalid")


def validate_guardian_email(email: str):
    try:
        validate_email(email)
    except ValidationError:
        raise InvalidEmailFormatError("Invalid email format")


def validate_guardian_data(guardian_data: dict):
    if "email" in guardian_data:
        validate_guardian_email(guardian_data["email"])
    return guardian_data
