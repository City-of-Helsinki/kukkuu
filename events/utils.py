import logging
from collections.abc import Iterable
from datetime import datetime
from typing import List, Optional, TYPE_CHECKING, Union

from django.conf import settings
from django.db.models import QuerySet
from django.utils import timezone
from django_ilmoitin.utils import send_notification
from parler.utils.context import switch_language

from common.utils import get_global_id
from users.utils import get_communication_unsubscribe_ui_url

if TYPE_CHECKING:
    from children.models import Child
    from events.consts import NotificationType
    from events.models import Event, EventGroup

logger = logging.getLogger(__name__)


def send_event_notifications_to_guardians(
    event: "Event",
    notification_type: "NotificationType",
    children: Union[QuerySet, List["Child"]],
    attachments: Optional[List] = None,
    **kwargs,
):
    if not isinstance(children, Iterable):
        children = [children]

    context_for_address = {}

    for child in children:
        for guardian in child.guardians.all():
            with switch_language(event, guardian.language):
                context = {
                    "event": event,
                    "child": child,
                    "guardian": guardian,
                    "event_url": get_event_ui_url(event, child, guardian.language),
                    "localtime": timezone.template_localtime,
                    "get_global_id": get_global_id,
                    "unsubscribe_url": get_communication_unsubscribe_ui_url(
                        guardian, guardian.language
                    ),
                    "is_obsolete": child.is_obsolete,
                    **kwargs,
                }
                occurrence = kwargs.get("occurrence")
                if occurrence:
                    context["occurrence_url"] = get_occurrence_ui_url(
                        occurrence, child, guardian.language
                    )
                    context["occurrence_enrol_url"] = get_occurrence_enrol_ui_url(
                        occurrence, child, guardian.language
                    )
                context_for_address[guardian.email] = context

    emails = ",".join(context_for_address.keys())
    logger.debug(
        f"There are {len(emails)} total in the notification list. "
        f"Notification will be sent to these addresses: {emails}"
    )
    # NOTE: Instead of sending the notification in the previous loop,
    # lets first collect the mail contexts to a dictionary and
    # send the mails after all the contexts have been handled properly.
    # This way we can prevent flooding the recipients with the mails
    # when some of the contexts cannot be handled because of any reason.
    # For more details, see
    # 1. https://helsinkisolutionoffice.atlassian.net/browse/KK-984,
    # 2. https://helsinkisolutionoffice.atlassian.net/browse/PT-1414.
    for address, context in context_for_address.items():
        logger.debug(f"Sending event notification to {address}.")
        guardian = context["guardian"]
        send_notification(
            email=address,
            notification_type=notification_type,
            context=context,
            language=guardian.language,
            attachments=attachments,
        )


def send_event_group_notifications_to_guardians(
    event_group: "EventGroup",
    notification_type: "NotificationType",
    children: Union[QuerySet, List["Child"]],
    **kwargs,
):
    if not isinstance(children, Iterable):
        children = [children]

    for child in children:
        for guardian in child.guardians.all():
            with switch_language(event_group, guardian.language):
                context = {
                    "event_group": event_group,
                    "child": child,
                    "guardian": guardian,
                    "event_group_url": get_event_group_ui_url(
                        event_group, child, guardian.language
                    ),
                    "localtime": timezone.template_localtime,
                    "get_global_id": get_global_id,
                    "events": [
                        {
                            "obj": event,
                            "event_url": get_event_ui_url(
                                event, child, guardian.language
                            ),
                        }
                        for event in event_group.events.language(guardian.language)
                    ],
                    "unsubscribe_url": get_communication_unsubscribe_ui_url(
                        guardian, guardian.language
                    ),
                    "is_obsolete": child.is_obsolete,
                    **kwargs,
                }

                send_notification(
                    guardian.email,
                    notification_type,
                    context=context,
                    language=guardian.language,
                )


def convert_to_localtime_tz(value):
    dt = datetime.combine(datetime.now().date(), value)
    if timezone.is_naive(value):
        # Auto add local timezone to naive time
        return timezone.make_aware(dt).timetz()
    else:
        return timezone.localtime(dt).timetz()


def get_event_ui_url(event, child, language):
    return "{}/{}/profile/child/{}/event/{}".format(
        settings.KUKKUU_UI_BASE_URL,
        language,
        get_global_id(child),
        get_global_id(event),
    )


def get_event_group_ui_url(event_group, child, language):
    return "{}/{}/profile/child/{}/event-group/{}".format(
        settings.KUKKUU_UI_BASE_URL,
        language,
        get_global_id(child),
        get_global_id(event_group),
    )


def get_occurrence_ui_url(occurrence, child, language):
    return "{}/{}/profile/child/{}/occurrence/{}".format(
        settings.KUKKUU_UI_BASE_URL,
        language,
        get_global_id(child),
        get_global_id(occurrence),
    )


def get_occurrence_enrol_ui_url(occurrence, child, language):
    return "{}/{}/profile/child/{}/event/{}/occurrence/{}/enrol".format(
        settings.KUKKUU_UI_BASE_URL,
        language,
        get_global_id(child),
        get_global_id(occurrence.event),
        get_global_id(occurrence),
    )
