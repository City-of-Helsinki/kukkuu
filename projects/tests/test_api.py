import pytest
from graphql_relay import to_global_id

from common.tests.utils import assert_permission_denied


@pytest.fixture(autouse=True)
def autouse_db(db):
    pass


PROJECTS_QUERY = """
query Projects {
  projects {
    edges {
      node {
        id
        name
        year
        translations {
          languageCode
          name
        }
      }
    }
  }
}
"""

PROJECT_QUERY = """
query Project($id: ID!) {
  project(id: $id){
    id
    name
    year
    translations {
      languageCode
      name
    }
  }
}
"""


def test_projects_query_unauthenticated(api_client, project):
    executed = api_client.execute(PROJECTS_QUERY)

    assert_permission_denied(executed)


def test_project_query_unauthenticated(snapshot, api_client, project):
    variables = {"id": to_global_id("ProjectNode", project.id)}

    executed = api_client.execute(PROJECT_QUERY, variables=variables)

    assert_permission_denied(executed)


def test_projects_query_normal_user(snapshot, guardian_api_client, project):
    executed = guardian_api_client.execute(PROJECTS_QUERY)

    snapshot.assert_match(executed)


def test_project_query_normal_user(snapshot, guardian_api_client, project):
    variables = {"id": to_global_id("ProjectNode", project.id)}

    executed = guardian_api_client.execute(PROJECT_QUERY, variables=variables)

    snapshot.assert_match(executed)
