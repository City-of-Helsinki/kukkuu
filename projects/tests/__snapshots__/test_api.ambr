# serializer version: 1
# name: test_project_query_normal_user
  dict({
    'data': dict({
      'project': dict({
        'enrolmentLimit': 2,
        'id': 'UHJvamVjdE5vZGU6MQ==',
        'name': 'Testiprojekti',
        'singleEventsAllowed': True,
        'translations': list([
          dict({
            'languageCode': 'FI',
            'name': 'Testiprojekti',
          }),
        ]),
        'year': 2020,
      }),
    }),
  })
# ---
# name: test_projects_query_normal_user
  dict({
    'data': dict({
      'projects': dict({
        'edges': list([
          dict({
            'node': dict({
              'enrolmentLimit': 2,
              'id': 'UHJvamVjdE5vZGU6MQ==',
              'name': 'Helsingin kaupunginorkesteri',
              'singleEventsAllowed': True,
              'translations': list([
                dict({
                  'languageCode': 'FI',
                  'name': 'Helsingin kaupunginorkesteri',
                }),
                dict({
                  'languageCode': 'SV',
                  'name': 'Helsingfors stadsorkester',
                }),
                dict({
                  'languageCode': 'EN',
                  'name': 'Helsinki Philharmonic Orchestra',
                }),
              ]),
              'year': 2020,
            }),
          }),
          dict({
            'node': dict({
              'enrolmentLimit': 2,
              'id': 'UHJvamVjdE5vZGU6Mg==',
              'name': 'Helsingin Kaupunginteatteri, Suomen Kansallisteatteri, Svenska Teatern, Teatteri ILMI Ö., Q-teatteri ja Nukketeatteri Sampo',
              'singleEventsAllowed': True,
              'translations': list([
                dict({
                  'languageCode': 'EN',
                  'name': 'Helsinki City Theatre, Finnish National Theatre, Svenska Teatern, Theatre ILMI Ö., Q-teatteri and Puppet Theatre Sampo',
                }),
                dict({
                  'languageCode': 'FI',
                  'name': 'Helsingin Kaupunginteatteri, Suomen Kansallisteatteri, Svenska Teatern, Teatteri ILMI Ö., Q-teatteri ja Nukketeatteri Sampo',
                }),
                dict({
                  'languageCode': 'SV',
                  'name': 'Helsingfors Stadsteater, Finlands Nationalteater, Svenska Teatern, Teatteri ILMI Ö., Q-teatteri och Dockteatern Sampo',
                }),
              ]),
              'year': 2021,
            }),
          }),
          dict({
            'node': dict({
              'enrolmentLimit': 2,
              'id': 'UHJvamVjdE5vZGU6Mw==',
              'name': 'Cirko - uuden sirkuksen keskus, Hotelli- ja ravintolamuseo, Suomen valokuvataiteen museo, Tanssin talo, Tanssiteatteri Hurjaruuth ja Teatterimuseo',
              'singleEventsAllowed': True,
              'translations': list([
                dict({
                  'languageCode': 'EN',
                  'name': 'Cirko – Center for New Circus, Hotel and Restaurant Museum, Finnish Museum of Photography, Dance House Helsinki, Dance Theatre Hurjaruuth and Theatre Museum',
                }),
                dict({
                  'languageCode': 'FI',
                  'name': 'Cirko - uuden sirkuksen keskus, Hotelli- ja ravintolamuseo, Suomen valokuvataiteen museo, Tanssin talo, Tanssiteatteri Hurjaruuth ja Teatterimuseo',
                }),
                dict({
                  'languageCode': 'SV',
                  'name': 'Cirko - Centrumet för nycirkus, Hotell- och restaurangmuseet, Dansens hus Helsingfors, Finlands fotografiska museum, Dansteatern Hurjaruuth och Teatermuseet',
                }),
              ]),
              'year': 2022,
            }),
          }),
          dict({
            'node': dict({
              'enrolmentLimit': 2,
              'id': 'UHJvamVjdE5vZGU6NA==',
              'name': 'Arkkitehtuurimuseo, Designmuseo, Helsingin kaupunginmuseo, Suomen kansallismuseo, Suomen kulttuuriperintökasvatuksen seura ja Tiedemuseo Liekki',
              'singleEventsAllowed': True,
              'translations': list([
                dict({
                  'languageCode': 'EN',
                  'name': 'Museum of Finnish Architecture, Design Museum, Helsinki City Museum, National Museum of Finland, Association of Cultural Heritage Education in Finland and Helsinki University Museum Flame',
                }),
                dict({
                  'languageCode': 'FI',
                  'name': 'Arkkitehtuurimuseo, Designmuseo, Helsingin kaupunginmuseo, Suomen kansallismuseo, Suomen kulttuuriperintökasvatuksen seura ja Tiedemuseo Liekki',
                }),
                dict({
                  'languageCode': 'SV',
                  'name': 'Finlands Arkitekturmuseum, Designmuseet, Helsingfors stadsmuseum, Finlands Nationalmuseum, Föreningen för kulturarvsfostran i Finland och Vetenskapsmuseet Lågan',
                }),
              ]),
              'year': 2023,
            }),
          }),
          dict({
            'node': dict({
              'enrolmentLimit': 2,
              'id': 'UHJvamVjdE5vZGU6NQ==',
              'name': 'Helsingin kirjasto- ja liikuntapalvelut',
              'singleEventsAllowed': True,
              'translations': list([
                dict({
                  'languageCode': 'EN',
                  'name': 'Helsinki City Library and Helsinki City Sports Services',
                }),
                dict({
                  'languageCode': 'FI',
                  'name': 'Helsingin kirjasto- ja liikuntapalvelut',
                }),
                dict({
                  'languageCode': 'SV',
                  'name': 'Helsingfors biblioteks– och idrottstjänster',
                }),
              ]),
              'year': 2024,
            }),
          }),
          dict({
            'node': dict({
              'enrolmentLimit': 2,
              'id': 'UHJvamVjdE5vZGU6Ng==',
              'name': 'HAM Helsingin taidemuseo, Amos Rex, Sointi Jazz Orchestra, UMO Helsinki Jazz Orchestra ja Kansallisgalleria: Ateneumin taidemuseo, Nykytaiteen museo Kiasma ja Sinebrychoffin taidemuseo',
              'singleEventsAllowed': True,
              'translations': list([
                dict({
                  'languageCode': 'EN',
                  'name': 'HAM Helsinki Art Museum, Amos Rex, Sointi Jazz Orchestra, UMO Helsinki Jazz Orchestra, Ateneum Art Museum, Museum of Contemporary Art Kiasma and Sinebrychoff Art Museum',
                }),
                dict({
                  'languageCode': 'FI',
                  'name': 'HAM Helsingin taidemuseo, Amos Rex, Sointi Jazz Orchestra, UMO Helsinki Jazz Orchestra ja Kansallisgalleria: Ateneumin taidemuseo, Nykytaiteen museo Kiasma ja Sinebrychoffin taidemuseo',
                }),
                dict({
                  'languageCode': 'SV',
                  'name': 'HAM Helsingfors konstmuseum, Amos Rex, Sointi Jazz Orchestra, UMO Helsinki Jazz Orchestra och Nationalgalleriet dvs. Ateneum, Kiasma och Konstmuseet Sinebrychoff',
                }),
              ]),
              'year': 2025,
            }),
          }),
        ]),
      }),
    }),
  })
# ---
